# Proceso de Limpieza de Datos
### Data Cleaning Process

---

__Dataset:__ Books Dataset  
__Fuente:__ https://www.kaggle.com/datasets/elvinrustam/books-dataset

A continuación se realiza una revisión completa de las características del dataset, y se procede implementar algunas modificaciones usuales del proceso de limpieza de datos (manejo de valores faltantes, entradas nulas, estandarización). El objetivo de este proyecto es genearar un archivo CSV con el dataset en óptimas condiciones para posteriores tareas, tales como el entrenamiento de un modelo predictivo o una visualización.

---

The following is a comprehensive review of the dataset's characteristics, followed by the implementation of standard data cleaning procedures (handling missing values, null entries, and standardization). The objective of this project is to generate a CSV file with the dataset in optimal condition for subsequent tasks, such as training a predictive model or creating visualizations.

In [3]:
# Parser: para transformación de fechas
# Parser: for date transformation 
from dateutil import parser 
import pandas as pd
import numpy as np

## 1. Carga y Exploración del dataset 
**Loading and Exploring the Dataset**


In [5]:
# Carga de datos y creación del dataframe
# Loading data and dataFrame setup
df = pd.read_csv("books-dataset.csv" , sep=",")
df.head()

,Title,Authors,Description,Category,Publisher,Publish Date,Price
0,Goat Brothers,"By Colton, Larry",NaN,"History , General",Doubleday,"Friday, January 1, 1993",Price Starting at $8.79
1,The Missing Person,"By Grumbach, Doris",NaN,"Fiction , General",Putnam Pub Group,"Sunday, March 1, 1981",Price Starting at $4.99
2,Don't Eat Your Heart Out Cookbook,"By Piscatella, Joseph C.",NaN,"Cooking , Reference",Workman Pub Co,"Thursday, September 1, 1983",Price Starting at $4.99
3,When Your Corporate Umbrella Begins to Leak: A...,"By Davis, Paul D.",NaN,NaN,Natl Pr Books,"Monday, April 1, 1991",Price Starting at $4.99
4,Amy Spangler's Breastfeeding : A Parent's Guide,"By Spangler, Amy",NaN,NaN,Amy Spangler,"Saturday, February 1, 1997",Price Starting at $5.32


In [7]:
# Resumen de los atributos del dataframe
# Overview of dataset attributes
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103082 entries, 0 to 103081
Data columns (total 7 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   Title         103082 non-null  object
 1   Authors       103082 non-null  object
 2   Description   70213 non-null   object
 3   Category      76912 non-null   object
 4   Publisher     103074 non-null  object
 5   Publish Date  103082 non-null  object
 6   Price         103082 non-null  object
dtypes: object(7)
memory usage: 5.5+ MB


**Observación 1:**

- El dataset cuenta con **7 campos/columnas** y un total de **103082 entradas**.
- En los campos _Description, Category_ y _Publisher_ se observa una cantidad de entradas faltantes.
- Por la naturaleza del dataset, la ausencia de los datos puede deberse a una causa en la fuente original (e.g., web librería, catálogo tienda).

**Remark 1:**

- The dataset consists of **7 fields/columns** and a total of **103,082 entries**.
- A number of missing entries are observed in the _Description, Category_, and _Publisher_ fields.
- Due to the nature of the dataset, the missing data may be attributed to issues within the original source (e.g., bookstore website, store catalog).

In [10]:
# Detalle de los incidentes en el dataset
# Detailed data quality issues
total_registros = len(df)
registros_duplicados  = df.duplicated().sum()
valores_faltantes = df.isnull().sum().to_dict()

print("---- CHEQUEO DE CALIDAD ----")
print(f"Registros Totales: {total_registros}")
print(f"Registros Duplicados: {registros_duplicados}")

print("Resumen Valores Faltantes: ")
for key, value in valores_faltantes.items():
    print(f"  {key}: {value} ({np.round(value/total_registros*100)} %)")

---- CHEQUEO DE CALIDAD ----
Registros Totales: 103082
Registros Duplicados: 8
Resumen Valores Faltantes: 
  Title: 0 (0.0 %)
  Authors: 0 (0.0 %)
  Description: 32869 (32.0 %)
  Category: 26170 (25.0 %)
  Publisher: 8 (0.0 %)
  Publish Date: 0 (0.0 %)
  Price: 0 (0.0 %)


---

## 2. Manejo de Valores Duplicados
**Handling Duplicate Values**

In [13]:
# Eliminación de los registros duplicados
# Deleting duplicate entries
df = df.drop_duplicates(keep='first')
total_registros_clean = len(df)
print(f">Registros Eliminados: {total_registros - total_registros_clean}")

>Registros Eliminados: 8


**Observación 2:**

- La cantidad de registros eliminados coincide con el conteo relizado más arriba.

**Remark 2:**
- The number of records removed is consistent with the previous count.

---

## 3. Estandarización y Manejo de los NaN/NaT
**Standardization and Handling of NaN/NaT**

In [18]:
# ========== Handling Price Column ==========
# Notemos que las entradas del campo "Price" presentan el texto "Price Starting at"
# Note that the entries in the "Price" field contain the text "Price Starting at"
df["Price"].head(10)

0     Price Starting at $8.79
1     Price Starting at $4.99
2     Price Starting at $4.99
3     Price Starting at $4.99
4     Price Starting at $5.32
5     Price Starting at $6.06
6     Price Starting at $4.99
7    Price Starting at $19.96
8     Price Starting at $4.99
9    Price Starting at $12.98
Name: Price, dtype: object

In [ ]:
# To-do List: 
# 1) Verificar si las todos los registros presentan el texto "Price Starting at"
# 2) Eliminar texto y modificar nombre del campo por "Price Starting at" 

# To-do List:
# 1) Verify if all records contain the text "Price Starting at"
# 2) Remove the text and rename the field to "Price Starting at"

# Lista para verificar que cada entrada poseea el texto "Price Starting at"
# List to verify that each entry contains the text "Price Starting at"
texto = []

# Lista para almacenar los tipos de moneda registrados
# List to store the recorded currency types
moneda = [] 

# Lista para transformar los valores de string a float
# List to convert string values to floats
valores = [] 

# Los primeros 17 caracteres de cada entrada corresponden a "Price Starting at"
# The first 17 characters of each entry correspond to "Price Starting at"
for registro in df["Price"]:
    texto.append(registro[:17])
    moneda.append(registro[18])
    valores.append(float("".join(registro[19:].split(","))))

In [22]:
# Todos los registros poseen el texto "Price Starting at"
# All records contain the text "Price Starting at"
print(f">Todas los registros con 'Price Starting at': {len(texto) == len(df)}")

# Los precios se expresan solo en una moneda
# All prices are listed in one currency only
print(f">Monedas Registradas: {set(moneda)}")

>Todas los registros con 'Price Starting at': True
>Monedas Registradas: {'$'}


In [ ]:
# Se agrega la columna/campo "Price Starting At" usando la lista "valores"
# Adding the "Price Starting At" column using the "valores" list
df["Price Starting At"] = valores

# Se elimina la column/campo "Price"
# Removing the "Price" column
df.drop("Price", axis=1, inplace=True)

In [30]:
# ========== Handling NaN ==========
# Description: NaN -> Not available
df["Description"] = df["Description"].fillna("Not available")

# Category: NaN - > Unassigned
df["Category"] = df["Category"].fillna("Unassigned")

# Publisher: NaN - > Unknown
df["Publisher"] = df["Publisher"].fillna("Unknown")

In [32]:
df.isnull().sum().to_dict()

{'Title': 0,
 'Authors': 0,
 'Description': 0,
 'Category': 0,
 'Publisher': 0,
 'Publish Date': 0,
 'Price Starting At': 0}

**Observación 3:**

- Se verifica que no existen _NaNs_ en el dataframe tras la modificación.

**Remark 3:**

- No _NaNs_ remain in the dataframe after processing.

In [101]:
# ========== Handling Dates ==========
# Se realiza la siguiente transformación:
# "DayName, MonthName DayNumber, Year" -> YYYY-MM-DD

# Applying the following transformation:
# "DayName, MonthName DayNumber, Year" -> YYYY-MM-DD

def analisis_flexible(fecha_string):
    """
    Utiliza un analizador (parser) inteligente para interpretar 
    de manera flexible el formato de la fecha y ejecutar la 
    transformación al formato YYYY-MM-DD.
    """
    try:
        return parser.parse(fecha_string).date()
    except:
        return pd.NaT
        
# Se crea la columna "Published on" a partir de la aplicación de función anterior
# Creating the "Published on" column from the previously applied function
df["Published on"] = pd.to_datetime(df["Publish Date"].apply(analisis_flexible), errors="coerce")

In [103]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 103074 entries, 0 to 103081
Data columns (total 8 columns):
 #   Column             Non-Null Count   Dtype         
---  ------             --------------   -----         
 0   Title              103074 non-null  object        
 1   Authors            103074 non-null  object        
 2   Description        103074 non-null  object        
 3   Category           103074 non-null  object        
 4   Publisher          103074 non-null  object        
 5   Publish Date       103074 non-null  object        
 6   Price Starting At  103074 non-null  float64       
 7   Published on       103069 non-null  datetime64[ns]
dtypes: datetime64[ns](1), float64(1), object(6)
memory usage: 11.1+ MB


**Observación 4:**

- Tras el cambio de formato, en la columna "Published on" aparecen 5 registros nulos.


**Remark 4:**

- After reformatting, there are 5 null entries in the "Published on" column.

In [108]:
# Inspección visual de los registros nulos en "Published on"
# Visually inspecting null entries in "Published on"
print(df[df["Published on"].isna()])

                                                  Title  \
48017  Father Taylor. A Story of Missionary Beginnings.   
56836                                          Rubaiyat   
71736                   Watkins Vanilla Classic Recipes   
77657                            Legal Spectator & More   
95902                        PENNSYLVANIA DUTCH COOKING   

                     Authors    Description    Category  \
48017       By Harvey Ingham  Not available  Unassigned   
56836        By Omar Khayyam  Not available  Unassigned   
71736                 By n/a  Not available  Unassigned   
77657      By Jacob A. Stein  Not available  Unassigned   
95902  By DUTCHCRAFT COMPANY  Not available  Unassigned   

                                             Publisher  \
48017                                          unknown   
56836                      Rahway: The Mershon Company   
71736                 Publications International, Ltd.   
77657  Washington, D.C.: The Magazine Group Washington   


**Observación 5:**

- Se observa que los 5 registros nulos contienen la fecha _Friday, January 1, 9999_ lo que parece ser un valor por defecto ante la ausencia de la información correspondiente.
- En este caso conviene reemplazar los NaT en la columna "Published on" por la fecha **9999-01-01** como valor por defecto.

**Remark 5:**

- It is observed that the 5 null records contain the date _Friday, January 1, 9999_, which appears to be a default value used when the corresponding information is missing.
- In this case, it is advisable to replace the NaTs in the "Published on" column with the date **9999-01-01** as a default value.

In [120]:
# Se refina el rango de fechas para evitar errores nuevos
# Refining the date range to prevent new errors
df["Published on"] = df["Published on"].astype("datetime64[us]")

# Reemplazo de los NaT con la fecha 9999-01-01
# Replacing NaTs with the date 9999-01-01
df["Published on"] = df["Published on"].fillna(pd.Timestamp("9999-01-01"))

In [128]:
# Se verifica que la cantidad de NaT en "Published on" sea cero
# Checking for zero NaT values in the "Published on"
df["Published on"].isnull().sum()

0

In [132]:
# Se elimina la columna "Publish Date"
# Dropping the "Publish Date" column
df.drop(columns=["Publish Date"], inplace=True)

# Chequeo visual del estado final del dataset limpio
# Final visual inspection of the clean dataset
df.head(3)

,Title,Authors,Description,Category,Publisher,Price Starting At,Published on
0,Goat Brothers,"By Colton, Larry",Not available,"History , General",Doubleday,8.79,1993-01-01
1,The Missing Person,"By Grumbach, Doris",Not available,"Fiction , General",Putnam Pub Group,4.99,1981-03-01
2,Don't Eat Your Heart Out Cookbook,"By Piscatella, Joseph C.",Not available,"Cooking , Reference",Workman Pub Co,4.99,1983-09-01


In [136]:
# Se exporta el dataset limpio en formato csv
# Saving the clean dataset as a CSV file
df.to_csv("books-dataset-clean.csv", index=False, encoding="utf-8")

---